### DISCUSSION TITLE
[Proposal]: Value safe enum

### DISCUSSION THREAD

**Discussion by IAmTheCShark** [dotnet/csharplang#6009] [Language Ideas] [answered]
I read all i could find here regarding enum, but nothing seemed to fit.
So I am opening a discussion here to introduce a new enum type or mechanism to prevent enums from having undefined values.

E.g. right now, this is legal and does not throw any exception nor compiler errors, even in checked context.

    public enum Foo
    {
        Bar = 1
    }

    Foo f = (Foo) 7;

    f = 0; // even worse in my eyes

My wish would be two parts here:

- Compiler should enforce using enum members. No casting from any constant allowed (also no special 0 treatment)
- Runtime (IConvertible) should check if it converts from a defined value.

Using the 'checked' context for this would be a breaking change, but the keyword seems to fit, so what about

    public checked enum Foo { ... }

**Comment by TahirAhmadov** [comment, marked as answer]
https://github.com/dotnet/runtime/discussions/61966

**Comment by huoyaoyuan** [comment]
#3179
